core/trino-main/src/main/java/io/trino/sql/analyzer/JsonPathAnalyzer.java

```diff
@@ -0,0 +1,533 @@
+            return INTEGER;
+        }
+
+        @Override
+        protected Type visitMemberAccessor(MemberAccessor node, Void context)
+        {
+            process(node.getBase());
+            return null;
+        }
+
+        @Override
+        protected Type visitNamedVariable(NamedVariable node, Void context)
+        {
+            Type parameterType = parameterTypes.get(node.getName());
+            if (parameterType == null) {
+                // This condition might be caused by the unintuitive semantics:
+                // identifiers in JSON path are case-sensitive, while non-delimited identifiers in SQL are upper-cased.
+                // Hence, a function call like JSON_VALUE(x, 'lax $var.floor()` PASSING 2.5 AS var)
+                // is an error, since the variable name is "var", and the passed parameter name is "VAR".
+                // We try to identify such situation and produce an explanatory message.
+                Optional<String> similarName = parameterTypes.keySet().stream()
+                        .filter(name -> name.equalsIgnoreCase(node.getName()))
+                        .findFirst();
+                if (similarName.isPresent()) {
+                    throw semanticException(INVALID_PATH, pathNode, format("no value passed for parameter %s. Try quoting \"%s\" in the PASSING clause to match case", node.getName(), node.getName()));
+                }
+                throw semanticException(INVALID_PATH, pathNode, "no value passed for parameter " + node.getName());
+            }
+
+            if (parameterType.equals(JSON_2016)) {
+                jsonParameters.add(PathNodeRef.of(node));
+                return null;
+            }
+
+            // in case of a non-JSON named variable, the type cannot be recorded and used as the result type of the node
+            // this is because any incoming null value shall be transformed into a JSON null, which is out of the SQL type system.
+            // however, for any incoming non-null value, the type will be preserved.
+            return null;
+        }
+
+        @Override
+        protected Type visitPredicateCurrentItemVariable(PredicateCurrentItemVariable node, Void context)
+        {
+            return null;
+        }
+
+        @Override
+        protected Type visitSizeMethod(SizeMethod node, Void context)
+        {
+            process(node.getBase());
+            types.put(PathNodeRef.of(node), INTEGER);
+            return INTEGER;
+        }
+
+        @Override
+        protected Type visitSqlValueLiteral(SqlValueLiteral node, Void context)
+        {
+            Type type = literalAnalyzer.analyze(node.getValue(), Scope.create());
+            types.put(PathNodeRef.of(node), type);
+            return type;
+        }
+
+        @Override
+        protected Type visitTypeMethod(TypeMethod node, Void context)
+        {
+            process(node.getBase());
+            Type type = createVarcharType(27);
```

Comment:
why `27` - please add a comment. Or get rid of magic number if possible.

Reply:
I extracted the type as a constant with a comment.
The `type()` method returns values from a SQL standard-defined pool. The longest is 27 chars long.